Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java

```diff
@@ -95,17 +94,31 @@ public int size() {
         };
     }
 
-    public DafnySequence<T> concatenate(DafnySequence<T> other) {
+    public final DafnySequence<T> concatenate(DafnySequence<T> other) {
         assert other != null : "Precondition Violation";
-        Object[] newArray = new Object[this.length() + other.length()];
-        int i = 0;
-        for (T t : this) {
-            newArray[i++] = t;
-        }
-        for (T t : other) {
-            newArray[i++] = t;
+
+        if (this.length() == 0) {
```

Comment:
How about an `isEmpty` predicate in `DafnySequence`?

Reply:
Sure.